src/index.rs

```diff
@@ -1,132 +1,278 @@
+            Some(s) => s,
+            None => return Err(Error::DfaHasNoStartState),
+        };
+
+        let mut transitions: HashMap<StateId, HashMap<TokenId, StateId>> = HashMap::default();
+        let mut final_states: HashSet<StateId> = HashSet::default();
+
+        let mut seen: HashSet<AutomataStateId> = HashSet::from_iter([start_state]);
+        let mut next_states: Vec<AutomataStateId> = vec![start_state];
+
+        while let Some(current_state) = next_states.pop() {
+            if dfa.is_match_state(dfa.next_eoi_state(current_state)) {
+                final_states.insert(current_state.as_u32());
             }
 
-            if fsm_info.finals.contains(&start_state) && !token_ids_end_states.is_empty() {
-                let inner_map = states_to_token_subsets.entry(start_state).or_default();
-                inner_map.insert(eos_token_id, start_state);
+            'token_loop: for (token, ids) in vocabulary.tokens_to_ids().iter() {
+                if ids.contains(&eos_token_id) {
+                    continue;
+                }
+
+                let mut next_state = current_state;
+                for transition_byte in token {
+                    next_state = dfa.next_state(next_state, *transition_byte);
+                    if dfa.is_dead_state(next_state) || dfa.is_quit_state(next_state) {
+                        continue 'token_loop;
+                    }
+                }
+
+                let is_intermediate_state = !dfa.is_match_state(next_state);
+                let is_full_match_state = dfa.is_match_state(dfa.next_eoi_state(next_state));
+                if is_intermediate_state || is_full_match_state {
+                    for token_id in ids {
+                        transitions
+                            .entry(current_state.as_u32())
+                            .or_default()
+                            .insert(*token_id, next_state.as_u32());
+                    }
+                }
+                if !seen.contains(&next_state) {
+                    seen.insert(next_state);
+                    next_states.push(next_state);
+                }
             }
+        }
 
-            seen.insert(start_state);
+        // Populate `transitions` with mappings from `final_states` to `eos_token_id`
+        for &final_state in &final_states {
+            transitions
+                .entry(final_state)
+                .or_default()
+                .insert(eos_token_id, final_state);
         }
 
-        let is_valid = states_to_token_subsets
-            .values()
-            .flat_map(|token_id_end_states| token_id_end_states.values())
-            .any(|end_state| fsm_info.finals.contains(end_state));
+        // Check if there is at least one valid mapping
+        let is_valid = transitions.values().any(|mapping| {
+            mapping
+                .values()
+                .any(|end_state| final_states.contains(end_state))
+        });
```

Comment:
FIXME

This `is_valid` was a leftover, but above we're literally populating transitions with final states, which we're re-checking here, I don't see how this check ever be invalid, so I think should be removed. But do we need something else to check for validity and vocabulary sufficiency?

For example, here we have empty vocabulary with only `eos_token_id` defined, do we consider it valid and sufficient?
```
let regex = "0|[1-9]*";
let vocabulary = Vocabulary::new(4);
println!("{:?}", Index::new(regex, &vocabulary));

> Ok(Index { initial_state: 40, final_states: {40}, transitions: {40: {4: 40}}, eos_token_id: 4 })

````

Reply:
Ah, but this check wouldn't even catch that case, because it doesn't check for reachability, it only checks for presence in a hash-map, which to which we've just added final states, reachable or not...

This check, in its current form, is definitely pointless.

Reply:
```
$ regex-cli debug sparse dfa '(?=a)b'
failed to parse pattern 0 to AST: '(?=a)b': failed to parse pattern: regex parse error:
    (?=a)b
    ^^^
error: look-around, including look-ahead and look-behind, is not supported
```

Guess this case doesn't matter either

Reply:
You should be able to remove this check, yes.

Reply:
@414owen @umut-sahin

So, do you think `Index` example from example above is valid? (built from an empty vocabulary with only `eos_token_id` defined). It will allow to generate only eos token for any regex given.

Reply:
It is as it matches the empty string. But even if wasn't, this is only a problem with custom vocabularies. With pre-built vocabularies, this is not a problem (as they contain all bytes as tokens).

That said, it might be a good idea to have an assertion for that in `from_pretrained`.

Reply:
Okay, got your point, thanks for confirming!